### PR TITLE
Minimal service architecture example

### PR DESCRIPTION
**Pull request by arrooney**
The purpose of this PR is to review the basic design that will be extended to implement different [ECSS services](http://everyspec.com/ESA/download.php?spec=ECSS-E-70-41A.047794.pdf). The accompanying docs are provided in the README that will guide AlbertaSat developers to build off this example.

This PR implements a dummy 'set time' command that simply accepts a CSP packet, routes it to the service handler, and if the argument is a valid unix timestamp (some time between 2020 and 2030), it will set a global time variable.

##### This does not implement command responses, or any actual hardware interfaces.

when used with the ex2_ground_station, the interaction looks like this (top terminal is running this code, and bottom terminal is running the ex2_ground_station code):

image

The CSP interface used is [ZMQ](zeromq.org)

### CODE REVIEW

File: main.c [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+      switch (csp_conn_dport(conn)) {
+        case TC_TIME_MANAGEMENT_SERVICE:
+          err = xQueueSendToBack(service_queues.time_management_app_queue,
+                                 packet, NORMAL_TICKS_TO_WAIT);
+          if (err != pdPASS) {
+            printf("FAILED TO QUEUE MESSAGE");
+          }
+          csp_buffer_free(packet);
+          break;
+
```
Comment (arrooney) [inline main.c:142]: The idea is to keep the server running to maximize the servers 'available time'. Commands are put into a freertos queue, and the server can keep recieving packets.

**Comment by arrooney** [comment]
> Consider using rebase -i to maintain a clean git history. https://medium.com/nona-web/cleaning-up-commit-history-with-git-rebase-61dc41d2829c.
> https://about.gitlab.com/blog/2018/06/07/keeping-git-commit-history-clean/
> For example, there are several commits with Update README as the message. While in some ways its good to show your process, it's not necessary to show us all of the details of the sausage being made. Using a rebase -i to fixup up all these commits, a clean history would have only one commit with the final changes to the README with a commit message something like "README: Add command handling readme." It is good practice to put the 'component' that you changed in caps preceding the commit message to allow for easier debugging. For example, if I am an SME in component Z and I suddenly notice that tests are failing with component Z even though I haven't made any changes, I can look at the history and see any recent PR's and the components they affect. If the history looks like this:
> COMPONENT Y: Add blah blah
> COMPONENT X: Add blah blah
> And I know that of the two, only X interacts with Z. I know almost immediately to look at that commit/PR without having to look at any code at all.
> Naturally, the prerequisite is that each commit is as self contained to a single component as possible, and that the number of commits in the PR is not overwhelming (as it is here). I'll comment on which commits I think should be merged together/broken/recombined.
> 
> Regardless of the history, the code looks good! It is however hard to review 60 commits thoroughly. I am happy to meet and walk through the rebase cleanup process with you so that we can trim that down to something more manageable.
> 
> While I realize this is a lot to ask in a review, you have the rare opportunity (this being one of the first PR's) of setting a clean history precedent before moving forward. Having a clean history makes a repo and all of its changes much easier to track and understand. This would greatly benefit both the AlbertaSat team as well as anyone basing their work off the opensource repo.

Excellent advice: I was talking to somone today about standards for these things (commit messages and branch names etc..). Perhaps we'll put some time into developing some standards for this!

**Comment by StefanVidakovic** [comment]
@arrooney 
Ok, let me know if you want a run down in a meeting or something. I've been fortunate enough to work somewhere with a top tier git history standard. More than happy to share the knowledge. Obviously you don't have to adopt my git clean history religion but I shall proselytize nonetheless.

**Comment by arrooney** [comment]
@StefanVidakovic that would be great if we could chat albertasat git & DevOps, I'd be curious to hear your thoughts! Y'all got time on Thursday after work?

**Comment by StefanVidakovic** [comment]
@arrooney Thursday is a no go for me. Lets take this offline (slack or email) to reduce PR clutter.

File: main.c [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  for (;;) {
+    /* Process incoming packet */
+    if ((conn = csp_accept(sock, 10000)) == NULL) {
+      /* timeout */
+      continue;
+    }
+    while ((packet = csp_read(conn, 50)) != NULL) {
+      switch (csp_conn_dport(conn)) {
+        case TC_TIME_MANAGEMENT_SERVICE:
+          err = xQueueSendToBack(service_queues.time_management_app_queue,
```
Comment (arrooney) [inline main.c:135]: The idea is to keep the server running to maximize the servers 'available time'. Commands are put into a freertos queue, and the server can keep recieving packets.

**Comment by arrooney** [comment]
Closing for now. Did a general review with Dr. Hoover, there are no high-level concerns. We're currently building on top of this.

**Pull request closed without merging**